### PR TITLE
`Feature`: i18n (Translations)

### PR DESCRIPTION
**Pull request by Shirkanesi**
Added I18N support to the tool, includes German translations

~~We might need to figure out if there is a better location for the ResourceBundle than the src-folder. 
Sadly I couldn't figure out whether there is a dedicated resource-folder somewhere when dealing with eclipse-plugins :(~~
~Edit: Figured out how to add such a folder :)~

TODO:
- [x] ~Create constant for default language (or stick with sytems locale?)~
- [x] Load available locales dynamically from the settings
- [x] Search for remaining strings that should be translated

### CODE REVIEW

**Comment by Shirkanesi** [comment]
In order to not overcomplicate the feature a restart of eclipse is required to update the language.
Otherwise we would need to separate the creation of every button/label/... from the code their behaviour and layout is defined. This would be an ugly solution (in my opinion)

**Comment by dfuchss** [comment]
I would suggest to load the actual strings into one ore more static classes that hold constants for the current language (could be some lazily loaded constants). These constants can be used in the classes instead of using the keys in the UI code directly. I think that might be easier, since you deal with the language in dedicated locations :) But that's discussable.

**Comment by Shirkanesi** [comment]
I've put this feature on pause until we figure out a good way of handling the configurations for both editions. (https://github.com/kit-sdq/programming-lecture-eclipse-artemis/issues/196)

**Comment by Shirkanesi** [comment]
> I would suggest to load the actual strings into one ore more static classes that hold constants for the current language (could be some lazily loaded constants). These constants can be used in the classes instead of using the keys in the UI code directly. I think that might be easier, since you deal with the language in dedicated locations :) But that's discussable.

I don't realy like to use constants for this. But I can imagine using an enum for this:
```java
public enum Messages {
	
	TAB_GRADING("tabs.grading"),
	TAB_ASSESSMENT_COURSE("tabs.assessment.course"),
	TAB_ASSESSMENT_EXAM("tabs.assessment.exam"),
	
	SETTINGS_DESCRIPTION("settings.description"),
	SETTINGS_URL("settings.url"),
	SETTINGS_USERNAME("settings.username"),
	SETTINGS_PASSWORD("settings.password")
	;
	
	private final String resourceBundleKey;
	
	Messages(String resourceBundleKey) {
		this.resourceBundleKey = resourceBundleKey;
	}
	
	public String getString() {
		return ResourceBundleProvider.getResourceBundle().getString(this.resourceBundleKey);
	}
	
	public String format(Object... params) {
		return String.format(this.getString(), params);
	}

}
```

**Comment by dfuchss** [comment]
I would suggest something like this (without resource files)

Define an interface I18N or L18N (as these abbr. are well known).
These interfaces can contain methods for each string we might want to use. (might be long but will be more readable in the code)

e.g., `L18N.settingsConfigPath()` instead of `ResourceBundleProvider.getResourceBundle().getString("settings.grading.absoluteConfigPath")`

All of these methods should be default methods like:

```java
if(LANGUAGE_OVERRIDE.isPresent()) // Static attribute that can be set if we do not want to use English
    return LANGUAGE_OVERRIDE.methodName()
return "My English Text for Default";
```

Furthermore, we would need some setter for the Language Override.


This kind of impl would have the following benefits:

* No need for a bunch of string variables as keys :D 
* Default Methods would ensure that a new field defaults to english if not translated (overridden in e.g., `GermanLang implements L18N`
* Method calls would be much shorter and therefore more readable.

The interface could look like this:

```java

interface L18N {
    static Container<L18N> LANGUAGE_OVERRIDE = Container.of(CommonActivator.getCurrentLanguage());

   deafault String settingAbsolutePath(){ // Maybe also use "?" then the one liner might be good readable.
       if(!LANGUAGE_OVERRIDE.isPresent() || isThisDefaultMethod()) // This has to be figured out (I'm not sure here). We have to make sure that we do not get an endless loop here (if override is present & method not overridden)
            return "Config Path Settings:"
       return LANGUAGE_OVERRIDE.get().settingAbsolutePath(); 
    }

}
```



What is your opinion on that ?

**Comment by Shirkanesi** [comment]
lgtm. I'll work on it

**Comment by dfuchss** [comment]
Thanks. The difficult part will be the default handling (as commented) I'm looking forward for the impl of that :)

**Comment by Shirkanesi** [comment]
I think those entire if-statements can be replaced by simple polymorphism. Just define the return of the default value in the interface and create a class for every override. 
Then the correct value will be returned due to dynamic binding.
(Given the I18N-object is swapped based on the selected language)

**Comment by dfuchss** [comment]
Yes, you are right. The only remaining part would be the swapping of the object. And the shortcut to it should be not that long :)

**Comment by dfuchss** [comment]
I've overseen that this morning because in my mind the methods were also static :D

**Comment by Shirkanesi** [comment]
I actually would just create a class like this:
```java
public class LanguageOverrideContainer {

	private static I18N currentOverride = new DefaultLanguage(); // todo swap
	
	public static I18N LANGUAGE_OVERRIDE() {
		return currentOverride;
	}
	
}
```

and just `import static ...LANGUAGE_OVERRIDE` everywhere we need it.

**Comment by dfuchss** [comment]
You would still have to call LANGUAGE_OVERRIDE() because it's a method

**Comment by Shirkanesi** [comment]
> You would still have to call LANGUAGE_OVERRIDE() because it's a method

Yes. But `LANGUAGE_OVERRIDE().settingsLanguage()` is quite readable and prevents anoying `.get()` calls :D

**Comment by dfuchss** [comment]
Yes. Maybe simply call the Method L18N or I18N  :)

**Comment by dfuchss** [comment]
And the container simply LanguageSettings

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/preferences/ArtemisPreferencesPage.java [hunk 1]
```diff
@@ -38,19 +47,42 @@ public ArtemisPreferencesPage() {
 	public void createFieldEditors() {
 		var parent = this.getFieldEditorParent();
 
-		var artemisUrl = new StringFieldEditor(PreferenceConstants.ARTEMIS_URL, "Artemis URL: ", parent);
-		var artemisUser = new StringFieldEditor(PreferenceConstants.ARTEMIS_USER, "Artemis Username: ", parent);
+		var artemisUrl = new StringFieldEditor(I18N().settingsUrl(), I18N().settingsUrl(), parent);
+		var artemisUser = new StringFieldEditor(I18N().settingsUsername(), I18N().settingsUsername(), parent);
+		var artemisPassword = new StringFieldEditor(I18N().settingsPassword(), I18N().settingsPassword(), parent);
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/preferences/ArtemisPreferencesPage.java:52]: PrefConstants should not be part of I18N :)

**Review by dfuchss — CHANGES_REQUESTED** [review]

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+ * NOTE: This file MUST use UTF-8 encoding. Otherwise spotless will fail or
+ * umlauts will break.
+ * 
+ * @see I18N
+ */
+class GermanLanguage implements I18N {
+
+	@Override
+	public String settingsLargeTextBox() {
+		return "Benutze mehrzeile Textbox für benutzerdefinierte Abzüge";
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java:16]:
```suggestion
		return "Benutze mehrzeilige Textbox für benutzerdefinierte Abzüge";
```

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+	}
+
+	@Override
+	public String settingsTextWrapping() {
+		return "Automatischen Zeilenumbruch in mehrzeiliger Textbox aktivieren";
+	}
+
+	@Override
+	public String settingsDescription() {
+		return "Einstellungen für Artemis Grading";
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java:26]:
```suggestion
		return "Einstellungen für Artemis";
```

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	}
+
+	@Override
+	public String settingsAdvancedDescription() {
+		return "Erweiterte Einstellungen für Artemis Grading";
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java:31]:
```suggestion
		return "Erweiterte Einstellungen für Artemis";
```

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/I18N.java [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+	default String settingsLargeTextBox() {
+		return "Use larger multiline text box for custom penalties";
+	}
+
+	default String settingsTextWrapping() {
+		return "Allow text-wrapping for multiline text boxes";
+	}
+
+	default String settingsDescription() {
+		return "Set preferences for Artemis Grading";
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/I18N.java:24]:
```suggestion
		return "Set preferences for Artemis";
```

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/I18N.java [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+	}
+
+	default String settingsAdvancedDescription() {
+		return "Set advanced settings for Artemis Grading";
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/I18N.java:28]:
```suggestion
		return "Set advanced settings for Artemis";
```

**Review by dfuchss — CHANGES_REQUESTED** [review]
First look on the translations :)

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/preferences/PreferenceInitializer.java [hunk 7]
```diff
@@ -25,6 +25,7 @@ public void initializeDefaultPreferences() {
 		store.setDefault(PreferenceConstants.GRADING_VIEW_PREFERS_LARGE_PENALTY_TEXT_PATH, false);
 		store.setDefault(PreferenceConstants.GRADING_VIEW_PREFERS_TEXT_WRAPPING_IN_PENALTY_TEXT_PATH, false);
 		store.setDefault(PreferenceConstants.GENERAL_OVERRIDE_DEFAULT_PREFERENCES, true);
+		store.setDefault(PreferenceConstants.GENERAL_PREFERRED_LANGUAGE, "English");
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/preferences/PreferenceInitializer.java:28]: I would suggest to get this from the LanguageSettings.

**Review by dfuchss — CHANGES_REQUESTED** [review]

**Review by dfuchss — APPROVED** [review]
Lgtm

**Pull request merged**